tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2019-2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "execute.hpp"
+#include "parser.hpp"
+#include <uvwasi.h>
+#include <cassert>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+// Global state.
+// NOTE: we do not care about uvwasi_destroy(), because it only clears file descriptors (currently)
+// and we are a single-run tool. This may change in the future and should reevaluate.
+uvwasi_t state;
```

Comment:
You can wrap it with `std::unique_ptr`.

Reply:
How would that help?

Reply:
It will call `uvwasi_destroy()` at the program exit.

Reply:
How? Is there a way to override the destructor of `unique_ptr`? Is the "second arg" that?

Reply:
Yeah, don't see too much benefit, or at least I couldn't make it nice.

Reply:
This looks to work
```
std::unique_ptr<uvwasi_t, void(*)(uvwasi_t*)> state{new uvwasi_t, uvwasi_destroy};
```

Reply:
I used `decltype(&uvwasi_destroy)` for the signature, but disliked using the `new` keyword and the indirection this adds without any benefit.

Reply:
The benefit is that `uvwasi_destroy` is used and we can detect issues with it.

There is a memory leak in @gumb0's suggestions.

Reply:
```suggestion
uvwasi_t state;
std::unique_ptr<uvwasi_t, decltype(&uvwasi_destroy)> state_destroyer{&state, &uvwasi_destroy};
```